### PR TITLE
Add AnalyticWhenIncoming boundary conditions for GH

### PR DESCRIPTION
**Pull request by prayush**
## Proposed changes

In progress: toward a BC that
 * does nothing (leaves boundary values unaltered) when char speeds are positive,
 * imposese analytic solution at the boundary when char speeds are negative.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+       min(char_speeds.at(3))}};
+  return *std::min_element(min_speeds.begin(), min_speeds.end());
+}
+
+template <typename T, typename DataType>
+T set_bc_when_char_speed_is_negative(const T& rhs_char_dt_u,
+                                     const T& desired_bc_dt_u,
+                                     const DataType& char_speed_u) noexcept {
+  auto bc_dt_u = rhs_char_dt_u;
+  auto it1 = bc_dt_u.begin();
```
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp]: it

File: src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+  auto it2 = desired_bc_dt_u.begin();
```
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp]: it_desired

File: src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+      }
+    }
+  }
+  return bc_dt_u;
+}
+}  // namespace BoundaryConditions_detail
+/// \ingroup ActionsGroup
+/// \brief Packages data on external boundaries for calculating numerical flux.
+/// Computes contributions on the interior side from the volume, and imposes
+/// Dirichlet boundary conditions on the exterior side.
```
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp]: on the exterior side, except it does nothing if all characteristic speeds are outgoing.

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
Looks good to me! Some minor suggestions. @nilsdeppe, would you please take a look at this, to make sure @prayush and I didn't miss something, before we try using it?

@prayush could you please add a unit test for this BC with some system that has a boundary with only outgoing char speeds and a boundary with incoming char speeds

File: src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp (hunk 1 above)
Comment (prayush) [reply src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp]: renamed

File: src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp (hunk 2 above)
Comment (prayush) [reply src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp]: renamed

File: src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp (hunk 3 above)
Comment (prayush) [reply src/Evolution/Systems/GeneralizedHarmonic/SemiAnalyticBoundaryConditions.hpp]: reworded!
